crypto/src/merkle.rs

```diff
@@ -200,34 +198,50 @@ impl<T> Node<T> {
             .collect();
         crate::Hash::new(sum).typed()
     }
+
+    fn bfs(&self) -> Vec<&Node<T>> {
+        let mut nodes = vec![self];
+        Self::_bfs_traverse(&mut nodes);
+        nodes
+    }
+
+    fn _bfs_traverse(node_list: &mut Vec<&Node<T>>) {
+        if node_list.is_empty() {
+            return;
+        }
+        let mut next_list = node_list
+            .iter()
+            .flat_map(|x| x.children())
+            .collect::<Vec<_>>();
+        Self::_bfs_traverse(&mut next_list);
+        node_list.append(&mut next_list);
+    }
+
+    fn children(&self) -> Vec<&Node<T>> {
+        match self {
+            Node::Subtree(subtree) => vec![&*subtree.left, &*subtree.right],
+            _ => vec![],
+        }
+    }
 }
 
 impl<'itm, T> BreadthFirstIter<'itm, T> {
     fn new(root_node: &'itm Node<T>) -> Self {
-        BreadthFirstIter {
-            queue: vec![root_node],
-        }
+        let mut nodes = root_node.bfs();
+        nodes.reverse();
+        BreadthFirstIter { stack: nodes }
```

Comment:
I don't like how this isn't lazily evaluated. I am of opinion that iterators should be lazily evaluated, because sometimes iteration can finish before iterator is fully depleted.

Reply:
This is another drawback. Can we find an algorithm that would make iterator lazily evaluated?

Reply:
I think it should be possible

Reply:
I wouldn't be so eager to jump into making this lazily evaluated. I'd wait before we had profiling data.